presto-benchmark-runner/src/main/java/com/facebook/presto/benchmark/prestoaction/JdbcPrestoAction.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Override
+    public QueryStats execute(Statement statement)
+    {
+        return execute(statement, Optional.empty()).getQueryStats();
+    }
+
+    @Override
+    public <R> QueryResult<R> execute(Statement statement, ResultSetConverter<R> converter)
+    {
+        return execute(statement, Optional.of(converter));
+    }
+
+    private <R> QueryResult<R> execute(Statement statement, Optional<ResultSetConverter<R>> converter)
+    {
+        return networkRetry.run(
+                "presto-cluster-connection",
+                () -> executeOnce(statement, converter));
+    }
+
+    private <R> QueryResult<R> executeOnce(Statement statement, Optional<ResultSetConverter<R>> converter)
+    {
+        String query = formatSql(statement, Optional.empty());
+        ProgressMonitor progressMonitor = new ProgressMonitor();
+
+        try (PrestoConnection connection = getConnection()) {
+            try (java.sql.Statement jdbcStatement = connection.createStatement()) {
+                PrestoStatement prestoStatement = jdbcStatement.unwrap(PrestoStatement.class);
+                prestoStatement.setProgressMonitor(progressMonitor);
+
+                ImmutableList.Builder<R> rows = ImmutableList.builder();
+                if (converter.isPresent()) {
+                    try (ResultSet resultSet = jdbcStatement.executeQuery(query)) {
+                        while (resultSet.next()) {
+                            rows.add(converter.get().apply(resultSet));
+                        }
+                    }
+                }
+                else {
+                    boolean moreResults = jdbcStatement.execute(query);
+                    if (moreResults) {
+                        consumeResultSet(jdbcStatement.getResultSet());
+                    }
+                    do {
+                        moreResults = jdbcStatement.getMoreResults();
+                        if (moreResults) {
+                            consumeResultSet(jdbcStatement.getResultSet());
+                        }
+                    }
+                    while (moreResults || jdbcStatement.getUpdateCount() != -1);
+                }
+
+                checkState(progressMonitor.getLastQueryStats().isPresent(), "lastQueryStats is missing");
+                return new QueryResult<>(rows.build(), progressMonitor.getLastQueryStats().get());
+            }
+        }
+        catch (SQLException e) {
+            throw exceptionClassifier.createException(progressMonitor.getLastQueryStats(), e);
+        }
+    }
+
+    private void consumeResultSet(ResultSet resultSet)
+            throws SQLException
+    {
+        while (resultSet.next()) {
+            // Do nothing
+        }
+    }
+
+    private PrestoConnection getConnection()
+            throws SQLException
+    {
+        PrestoConnection connection = DriverManager.getConnection(jdbcUrl, "user", null)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getConnection*. Currently, there are execution paths that do not contain closure statements, for example, when Wrapper.unwrap throws an exception. Ensure the object returned by *getConnection()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).